File: 19-remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode fast=head;
        int count=0;
        while(count<n){
            fast=fast.next;
            count++;
        }

        if(fast==null){
            return head.next;
        }
     //   System.out.println("Fast: "+fast.val);
        ListNode slow=head;

        while(fast.next!=null){
            slow=slow.next;
            fast=fast.next;
        }

        slow.next=slow.next.next;

        return head;

    }
}

File: 19-remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };

 1. move fast pointer n steps
 2. slow=head
 3. move both slow & fast pointer 1 step => fast.next!=null
 4. slow.next=slow.next.next
 */
struct ListNode* removeNthFromEnd(struct ListNode* head, int n) {
    
    struct ListNode *fast=head;
    int count=1;

    while(count<=n){
        fast=fast->next;
        count++;
    }
    if(fast==NULL){
        return head->next;
    }
    struct ListNode *slow=head;
    while(fast->next!=NULL){
        slow=slow->next;
        fast=fast->next;
    }

    slow->next=slow->next->next;

    return head;
}
















